app/src/main/java/pm/gnosis/heimdall/ui/safe/recover/extension/ReplaceExtensionQrActivity.kt

```diff
@@ -0,0 +1,185 @@
+import pm.gnosis.heimdall.data.repositories.PushServiceRepository
+import pm.gnosis.heimdall.di.components.ViewComponent
+import pm.gnosis.heimdall.reporting.ScreenId
+import pm.gnosis.heimdall.ui.base.ViewModelActivity
+import pm.gnosis.heimdall.ui.qrscan.QRCodeScanActivity
+import pm.gnosis.heimdall.utils.handleQrCodeActivityResult
+import pm.gnosis.model.Solidity
+import pm.gnosis.svalinn.common.utils.shareExternalText
+import pm.gnosis.svalinn.common.utils.snackbar
+import pm.gnosis.svalinn.common.utils.toast
+import pm.gnosis.svalinn.common.utils.visible
+import pm.gnosis.utils.asEthereumAddress
+import pm.gnosis.utils.asEthereumAddressString
+import timber.log.Timber
+import javax.inject.Inject
+import kotlinx.android.synthetic.main.screen_replace_extension_qr.replace_extension_back_arrow as backArrow
+import kotlinx.android.synthetic.main.screen_replace_extension_qr.replace_extension_bottom_panel as bottomPanel
+import kotlinx.android.synthetic.main.screen_replace_extension_qr.replace_extension_coordinator as coordinator
+import kotlinx.android.synthetic.main.screen_replace_extension_qr.replace_extension_extension_link as extensionLink
+import kotlinx.android.synthetic.main.screen_replace_extension_qr.replace_extension_progress_bar as progressBar
+
+class ReplaceExtensionQrActivity : ViewModelActivity<ReplaceExtensionQrContract>() {
+
+    override fun layout() = R.layout.screen_replace_extension_qr
+
+    override fun inject(component: ViewComponent) = viewComponent().inject(this)
+
+    override fun screenId() = ScreenId.REPLACE_BROWSER_EXTENSION_QR
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        val safeAddress = (intent.extras[EXTRA_SAFE_ADDRESS] as String)?.asEthereumAddress()
+        safeAddress?.let {
+            viewModel.setup(it)
+        } ?: finish()
+
+
+        viewModel.observableState.observe(this, Observer {
+
+            onPairingLoading(it.isLoading)
+
+            it.pairingResult?.let {
+
+                when(it) {
+
+                    is ReplaceExtensionQrContract.PairingResult.PairingSuccess -> {
+                        toast(R.string.devices_paired_successfuly)
+                        startActivity(ReplaceExtensionRecoveryPhraseActivity.createIntent(this, safeAddress!!, it.extension))
+                    }
+
+                    is ReplaceExtensionQrContract.PairingResult.PairingError -> {
+                        snackbar(coordinator, R.string.error_pairing_devices)
+                        Timber.e(it.error)
+                    }
+                }
+            }
+        })
+
+        extensionLink.apply {
+            this.text = SpannableStringBuilder(Html.fromHtml(getString(R.string.pairing_first_step)))
+            setOnClickListener { shareExternalText(BuildConfig.CHROME_EXTENSION_URL, "Browser Extension URL") }
+        }
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+
+        backArrow.setOnClickListener {
```

Comment:
the extensionLink click listener is set in the onCreate, do we want to do all in the same place (doesn't really matter which one)?